change(web): leverage tokenization to preserve punctuation and whitespace when predicting 📚

jahorton opened the pull request:
This PR integrates the new transform-tokenization function from #11998 with the context-tokenization / caching engine in order to determine which parts of a transform affect old tokens vs affect the current token up for correction/prediction.

Combining its results with the reworked context-tokenization logic, we can now detect whenever the incoming text results in a new context token, regardless of source - be it whitespace, punctuation, or any other cause (as would be the case for #5025 - dictionary-based wordbreaking).  We are then able to recognize what part of the input, if any, belongs to "previous" tokens and thus should not be replaced by newly-generated predictions - something vital for improving our handling of prediction and auto-correct after newly-typed punctuation marks (in regard to #11963, list item 1).

## User Testing

Use the EuroLatin (SIL) keyboard and the standard MTNT model for the following tests.

**TEST_GENERAL_PREDICTION_USE**:  Thoroughly test the predictive-text feature for Keyman by using an appropriate mobile device.

**TEST_PREDICTION_ON_PUNCTUATION**:  Test the behavior of predictive-text when the most-recently typed text is an English punctuation mark - `.`, `,`, `-`, `?`, and/or `!`.
- Ideally, suggestion will not be displayed - English words don't exactly start with any of these characters.
- If they do, applying a suggestion should not affect any part of the word preceding it.

**TEST_PREDICTION_AFTER_HYPHEN**:  Verify that applying predictions for text after a hyphen acts as would be expected - the suggestions should only affect text after the hyphen.

**TEST_AUTO_PREDICT_AFTER_PUNCTUATION**:  Type two standard English sentences, using formal punctuation as appropriate.  Verify that punctuation remains unaffected by auto-predict.
- Do _not_ use single-quotes unless typing contractions.

**TEST_AUTO_PREDICT_CONTRACTIONS**: Type multiple standard English words, then start typing a contraction, stopping immediately after the `'` character.
- So, for `can't`, stop at `can'`
- Good cases can be found with:
    - `aren'`
    - `wouldn'`
- If the full contraction is auto-selected at this position, verify that pressing the spacebar auto-accepts it.
- Known issue:  if the text before the `'` is a standard English word, that word will likely be auto-selected, rather than the contraction.

Note:  typing punctuation marks will _not_ trigger auto-correct at this time.  We don't currently have support for that; such a feature will require extra design work, as different languages use different punctuation marks.

jahorton commented on common/web/lm-worker/src/main/correction/context-tracker.ts:
```diff
@@ -407,41 +373,81 @@ export class ContextTracker extends CircularArray<TrackedContextState> {
     // Reset priorEdit for the end-of-context updating loop.
     priorEdit = undefined;
 
+    // TODO:  add flag:  rebuiltSearchSpace (?)
+    //
+    // Or should searchSpace be tracked (eventually) on the tokens, rather than on the 'state'?
+    // - I'm leaning toward the latter here.
+    // - Shifting the paradigm should be a separate work unit than the context-tracker rework
+    //   currently being done, though.
```
These are notes toward potential future work - things that would need to be addressed to support phrase-level corrections.

jahorton commented on common/web/lm-worker/src/main/correction/context-tracker.ts:
```diff
@@ -397,10 +447,31 @@ export class ContextTracker extends CircularArray<TrackedContextState> {
           const token = state.tokens[i - poppedTokenCount];
           const matchToken = matchState.tokens[i];
 
+          // TODO:  I'm beginning to believe that searchSpace should (eventually) be tracked
+          // on the tokens, rather than on the overall 'state'.
+          // - Reason:  phrase-level corrections / predictions would likely need a search-state
+          //   across per potentially-affected token.
+          // - Shifting the paradigm should be a separate work unit than the
+          //   context-tracker rework currently being done, though.
```
These are notes toward potential future work - things that would need to be addressed to support phrase-level corrections.

jahorton replied:
Added as part of a new issue I'm writing up in relation to this - #12010.

dinakaranr commented:
# Test Results
I tested this issue with the attached "Keyman 18.0.75-alpha-test-12015" build on the Android 14 & iPhone. Here is my observation.

Prerequisite Steps:
1. Installed the "Keyman-18.0.75" build for Android 14 & iPhone 13 for the testflight app and gave all permissions to the application.
2. Checked the "Enable Keyman as system-wide keyboard" and set the keyboard as the default keyboard box on the settings page.
3. Open the Keyman app. Enable the "Predictions" and install the "Dictionary."
4. Installed the "EuroLatin (SIL)" keyboard(v 3.0.2) and the standard MTNT dictionary(v0.3.2) model.

* **TEST_GENERAL_PREDICTION_USE (Passed):**
1. I followed the prerequisite steps on Android 14 & iPhone 13. 
2. Added a sentence to the keyman notepad and chrome(search).
3. Enter a word and observe that the word is highlighted on the banner. 
4. Press the "Spacebar" and then the word added. 
The word is selected using a cursor from left to right and right to left. 
The word prediction works on the banner and selection works correctly.

* **TEST_PREDICTION_ON_PUNCTUATION (Passed):**
1. I followed the prerequisite steps on Android 14 & iPhone 13. 
2. Added a sentence to the keyman notepad and chrome(search).
3. Enter a word and then the suggestion does not start with any punctuation mark(-,.,?,!)
Suggestion works well then when typing with punctuation.

* **TEST_PREDICTION_AFTER_HYPHEN (Passed):** 
1. I followed the prerequisite steps on Android 14 & iphone 13. 
2. Added a sentence to the keyman notepad and chrome(search).
3. Enter a word and then the suggestion works after the hyphen.

 
* **TEST_AUTO_PREDICT_AFTER_PUNCTUATION (Failed):**
1. I followed the prerequisite steps on Android 14 & iphone 13. 
2. I added a sentence to the keyman notepad and Chrome (search).
3. Enter a word within double quotes. 
4. Move the cursor after double quotes. Here, the suggestion word is highlighted on the banner.

Press the spacebar and then double-quotes disappear with blank spaces. 
Add a double-quotes after the word. Here, a double-quote suggestion appears on the banner. 
I explained this behavior in the video. Please refer to it. It may be a problem. Here, I failed this scenario. Thanks.
https://github.com/user-attachments/assets/9d9bc0a9-0d12-41f5-87fe-915ce7bce983

jahorton commented:
> * **TEST_AUTO_PREDICT_AFTER_PUNCTUATION (Failed):**
> 
> 
>     1. I followed the prerequisite steps on Android 14 & iphone 13.
> 
>     2. I added a sentence to the keyman notepad and Chrome (search).
> 
>     3. Enter a word within double quotes.
> 
>     4. Move the cursor after double quotes. Here, the suggestion word is highlighted on the banner.
> 
> 
> Press the spacebar and then double-quotes disappear with blank spaces. Add a double-quotes after the word. Here, a double-quote suggestion appears on the banner. I explained this behavior in the video. Please refer to it. It may be a problem. Here, I failed this scenario. Thanks. https://github.com/user-attachments/assets/9d9bc0a9-0d12-41f5-87fe-915ce7bce983

See:  timestamp around 0:42.  (It would really help to mention the moment things go wrong in these reports.)

mcdurdin commented on common/web/lm-worker/src/main/correction/context-tracker.ts:
```diff
@@ -79,6 +80,18 @@ export class TrackedContextToken {
+    // Preserve existing text if new text isn't specified.
+    tokenText = tokenText || (tokenText === '' ? '' : this.raw);
+
+    if(transformDistribution && transformDistribution.length > 0) {
```
```suggestion
    if(transformDistribution?.length > 0) {
```

mcdurdin commented on common/web/lm-worker/src/main/predict-helpers.ts:
```diff
@@ -667,6 +645,22 @@ export function finalizeSuggestions(
+      let mergedTransform = models.buildMergedTransform(tuple.preservationTransform, prediction.sample.transform);
+      mergedTransform.id = prediction.sample.transformId;
+
+      // Temporarily and locally drops 'readonly' semantics so that we can reassign the transform.
+      // See https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html#improved-control-over-mapped-type-modifiers
+      let mutableSuggestion = prediction.sample as {-readonly [transform in keyof Suggestion]: Suggestion[transform]};
+
+      // Assignment via by-reference behavior, as suggestion is an object
+      mutableSuggestion.transform = mergedTransform;
```
This feels like a TODO for future cleanup? I appreciate it isn't new code...

mcdurdin reviewed (approved):
LGTM

jahorton commented:
> > * **TEST_AUTO_PREDICT_AFTER_PUNCTUATION (Failed):**
> >   
> >   1. I followed the prerequisite steps on Android 14 & iphone 13.
> >   2. I added a sentence to the keyman notepad and Chrome (search).
> >   3. Enter a word within double quotes.
> >   4. Move the cursor after double quotes. Here, the suggestion word is highlighted on the banner.
> > 
> > Press the spacebar and then double-quotes disappear with blank spaces. Add a double-quotes after the word. Here, a double-quote suggestion appears on the banner. I explained this behavior in the video. Please refer to it. It may be a problem. Here, I failed this scenario. Thanks. https://github.com/user-attachments/assets/9d9bc0a9-0d12-41f5-87fe-915ce7bce983
> 
> See: timestamp around 0:42. (It would really help to mention the timestamp where things go wrong in these video reports.)

Some followup:

If using the Web "Predictive Text: robust testing" test page, typing `""` and then moving the caret (via click) between the quotes, an error is produced:

```
tokenization.ts:144 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'isWhitespace')
    at tokenize (tokenization.ts:144:39)
    at getLastPreCaretToken (tokenization.ts:188:22)
    at _TrieModel.predict (trie-model.ts:316:18)
    at predictFromCorrections (predict-helpers.ts:373:36)
    at predict-helpers.ts:292:23
    at Generator.next (<anonymous>)
    at fulfilled (ef40fcaa-765c-4003-81e9-e14c0a0a1361:36:26)
```

Something like this occurs near the error's timestamp, though with a different total context.  I also remember having a similar issue at one point in development when _backspacing_ to a similar point, rather than directly moving the caret.

jahorton commented:
It turns out that we had some cases where the engine was trying to use corrections that erased the current token, causing predictions to be based on the previous text.  Since there aren't really any good matches for single punctuation-mark strings, the correction search was sometimes trying to perform such a correction and predict from the result... which is why the `"` was being ignored.  I'm still not sure why we didn't see it with `.`, but I believe I've fixed the issue.

To be safe and ensure this fix doesn't break something else while we're at it...

@keymanapp-test-bot retest all

dinakaranr commented:
# Test Results
I tested this issue with the attached "Keyman 18.0.60-alpha-test-11997" build on the Android 14 & iPhone 13. Here is my observation.

Prerequisite Steps:
Installed the "Keyman-18.0.60" build for Android 14 & iPhone 13 for the testflight app and gave all permissions to the application.
Checked the "Enable Keyman as system-wide keyboard" and set the keyboard as the default keyboard box on the settings page.
Open the Keyman app. Enable the "Predictions" and install the "Dictionary."
Installed the "EuroLatin (SIL)" keyboard(v 3.0.2) and the standard MTNT dictionary(v0.3.2) model.

* **TEST_GENERAL_PREDICTION_USE (passed):**

I followed the prerequisite steps on Android 14 & iPhone 13.
Added a sentence to the keyman notepad and chrome(search).
Enter a word and observe that the word is highlighted on the banner. 
Press the "Spacebar" and then the word added.
The word is selected using a cursor from left to right and right to left.
The word prediction works on the banner and selection works correctly.

* **TEST_PREDICTION_ON_PUNCTUATION (failed):**

I followed the prerequisite steps on Android 14 & iPhone 13.
Added a sentence to the keyman notepad and chrome(search).
Enter a sentence and end with a question. ('Is it right?')  suggestion appears on banner if cursor after the question.(-,.,?,!)
Press the spacebar. Here, the question disappears
Press the backspace. Here, the question icon appears on the banner within double quotes.
I explained this behavior in the video. Please refer to it. It may be a problem. Here, I failed this scenario. Thanks.
https://github.com/user-attachments/assets/d6f74325-fbf8-4c56-8c76-ea93c665487e

* **TEST_PREDICTION_AFTER_HYPHEN (passed):**

I followed the prerequisite steps on Android 14 & iphone 13.
Added a sentence to the keyman notepad and chrome(search).
Enter a word and then the suggestion works after the hyphen.

* **TEST_AUTO_PREDICT_AFTER_PUNCTUATION (failed):**

I followed the prerequisite steps on Android 14 & iphone 13.
I added a sentence to the keyman notepad and Chrome (search).
Enter a word within single quotes.
Move the cursor after single quotes. Here, the suggestion word is highlighted on the banner.
Press the spacebar and then single-quotes disappear with blank spaces.
Press the backspace. Here, a double-quote suggestion appears on the banner. but It is showing the word along with single quotes
I explained this behavior in the video. Please refer to it. It may be a problem. Here, I failed this scenario. Thanks.
https://github.com/user-attachments/assets/a5c4c8c6-ab04-48e3-a418-5af5174b65ad
I am sorry to fail these scenarios.

jahorton commented:
> * **TEST_PREDICTION_ON_PUNCTUATION (failed):**
> 
> 
> I followed the prerequisite steps on Android 14 & iPhone 13. Added a sentence to the keyman notepad and chrome(search). Enter a sentence and end with a question. ('Is it right?') suggestion appears on banner if cursor after the question.(-,.,?,!) Press the spacebar. Here, the question disappears Press the backspace. Here, the question icon appears on the banner within double quotes. I explained this behavior in the video. Please refer to it. It may be a problem. Here, I failed this scenario. Thanks. https://github.com/user-attachments/assets/d6f74325-fbf8-4c56-8c76-ea93c665487e

Timestamp:  ~ 0:24.

After:
1. starting a new sentence + initial non-whitespace token, 
2. then backspacing past the sentence-ending punctuation mark,
3. and _then_ retyping the deleted punctuation mark

It continued to predict as if the punctuation mark was part of the token it follows.



> * **TEST_AUTO_PREDICT_AFTER_PUNCTUATION (failed):**
> 
> 
> I followed the prerequisite steps on Android 14 & iphone 13. I added a sentence to the keyman notepad and Chrome (search). Enter a word within single quotes. Move the cursor after single quotes. Here, the suggestion word is highlighted on the banner. Press the spacebar and then single-quotes disappear with blank spaces. Press the backspace. Here, a double-quote suggestion appears on the banner. but It is showing the word along with single quotes I explained this behavior in the video. Please refer to it. It may be a problem. Here, I failed this scenario. Thanks. https://github.com/user-attachments/assets/a5c4c8c6-ab04-48e3-a418-5af5174b65ad I am sorry to fail these scenarios.

I understand the confusion here, but this test **PASSED**.  We need special handling for English contractions - if the word there had been `can'`, the user may have been typing `can't` and stopped - we want the model to be able to suggest `can't` and for it to auto-apply when the situation arises.  This is _specific_ to cases with single-quotes, though.

jahorton commented:
I believe I've corrected the issues previously seen.  Part of this includes defining a new unit test so that the special behaviors involving `'` are more explicit.

@keymanapp-test-bot retest all

keyman-server commented:
Changes in this pull request will be available for download in [Keyman version 18.0.83-alpha](https://keyman.com/downloads/releases/alpha/18.0.83)
